File: ios/Plugin/NativeJobEvaluator.swift
import Foundation

struct NativeJobEvent {
    let id: String
    let name: String
    let firedAt: Int64
    let source: String
    let data: [String: Any]?
}

final class NativeJobEvaluator {
    private static let storageKey = "mobilecron:state"
    private static let clockRegex = try? NSRegularExpression(pattern: #"^(\d{2}):(\d{2})$"#)

    // ── Shared I/O ─────────────────────────────────────────────────────────────

    /// JSON file in Application Support — survives simctl terminate (synchronous atomic write).
    static var stateFileURL: URL? {
        guard let dir = try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        ) else { return nil }
        return dir.appendingPathComponent("mobilecron-state.json")
    }

    /// Reads raw JSON: file preferred (reliable); UserDefaults as fallback.
    static func readStateRaw() -> String? {
        if let url = stateFileURL,
           let raw = try? String(contentsOf: url, encoding: .utf8),
           !raw.isEmpty {
            return raw
        }
        return UserDefaults.standard.string(forKey: storageKey)
    }

    /// Writes raw JSON atomically to file AND to UserDefaults (belt-and-suspenders).
    static func writeStateRaw(_ raw: String) {
        if let url = stateFileURL {
            try? raw.write(to: url, atomically: true, encoding: .utf8)
        }
        UserDefaults.standard.set(raw, forKey: storageKey)
    }

    // ── Evaluation ─────────────────────────────────────────────────────────────

    static func evaluate(source: String) -> [NativeJobEvent] {
        guard let raw = readStateRaw(),
              let data = raw.data(using: .utf8),
              var state = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return []
        }

        var jobs = state["jobs"] as? [[String: Any]] ?? []
        var pendingEvents = state["pendingNativeEvents"] as? [[String: Any]] ?? []
        var firedEvents: [NativeJobEvent] = []
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let paused = bool(from: state["paused"]) ?? false
        var mutated = false

        for index in jobs.indices {
            var job = jobs[index]
            defer { jobs[index] = job }

            guard (bool(from: job["enabled"]) ?? false) else { continue }

            if int64(from: job["nextDueAt"]) == nil,
               let computed = computeNextDueAt(schedule: job["schedule"] as? [String: Any], nowMs: now) {
                job["nextDueAt"] = computed
                mutated = true
            }

            guard let nextDueAt = int64(from: job["nextDueAt"]), nextDueAt <= now else { continue }

            if getSkipReason(job: job, paused: paused, nowMs: now) != nil {
                job["consecutiveSkips"] = (int64(from: job["consecutiveSkips"]) ?? 0) + 1
                job["updatedAt"] = now
                if scheduleKind(job["schedule"] as? [String: Any]) == "every" {
                    if let next = computeNextDueAt(schedule: job["schedule"] as? [String: Any], nowMs: now) {
                        job["nextDueAt"] = next
                    } else {
                        job.removeValue(forKey: "nextDueAt")
                    }
                }
                mutated = true
                continue
            }

            job["lastFiredAt"] = now
            job["updatedAt"] = now
            job["consecutiveSkips"] = 0

            var eventPayload: [String: Any] = [
                "id": (job["id"] as? String) ?? "",
                "name": (job["name"] as? String) ?? "",
                "firedAt": now,
                "source": source
            ]
            if let eventData = job["data"] as? [String: Any] {
                eventPayload["data"] = eventData
            }

            if scheduleKind(job["schedule"] as? [String: Any]) == "at" {
                job["enabled"] = false
                job.removeValue(forKey: "nextDueAt")
            } else {
                if let next = computeNextDueAt(schedule: job["schedule"] as? [String: Any], nowMs: now) {
                    job["nextDueAt"] = next
                } else {
                    job.removeValue(forKey: "nextDueAt")
                }
            }

            pendingEvents.append(eventPayload)
            firedEvents.append(
                NativeJobEvent(
                    id: eventPayload["id"] as? String ?? "",
                    name: eventPayload["name"] as? String ?? "",
                    firedAt: int64(from: eventPayload["firedAt"]) ?? now,
                    source: eventPayload["source"] as? String ?? source,
                    data: eventPayload["data"] as? [String: Any]
                )
            )
            mutated = true
        }

        if mutated {
            state["jobs"] = jobs
            if pendingEvents.isEmpty {
                state.removeValue(forKey: "pendingNativeEvents")
            } else {
                state["pendingNativeEvents"] = pendingEvents
            }

            if let nextData = try? JSONSerialization.data(withJSONObject: state),
               let nextRaw = String(data: nextData, encoding: .utf8) {
                writeStateRaw(nextRaw)
            }
        }

        return firedEvents
    }

    private static func computeNextDueAt(schedule: [String: Any]?, nowMs: Int64) -> Int64? {
        guard let schedule else { return nil }
        switch scheduleKind(schedule) {
        case "at":
            guard let atMs = int64(from: schedule["atMs"]) else { return nil }
            return atMs > nowMs ? atMs : nil
        case "every":
            guard let everyMs = int64(from: schedule["everyMs"]), everyMs > 0 else { return nil }
            let anchorMs = int64(from: schedule["anchorMs"]) ?? nowMs
            if nowMs < anchorMs { return anchorMs }
            let elapsed = nowMs - anchorMs
            let steps = (elapsed / everyMs) + 1
            return anchorMs + (steps * everyMs)
        default:
            return nil
        }
    }

    private static func getSkipReason(job: [String: Any], paused: Bool, nowMs: Int64) -> String? {
        if paused { return "paused" }
        if let activeHours = job["activeHours"] as? [String: Any],
           !isWithinActiveHours(activeHours: activeHours, nowMs: nowMs) {
            return "outside_active_hours"
        }

        // BGTask constraints should enforce network/charging requirements on iOS.
        return nil
    }

    private static func isWithinActiveHours(activeHours: [String: Any], nowMs: Int64) -> Bool {
        guard let startValue = activeHours["start"] as? String,
              let endValue = activeHours["end"] as? String,
              let start = parseClock(startValue),
              let end = parseClock(endValue) else {
            return true
        }

        let timeZone: TimeZone
        if let tzId = activeHours["tz"] as? String, let tz = TimeZone(identifier: tzId) {
            timeZone = tz
        } else {
            timeZone = .current
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let nowDate = Date(timeIntervalSince1970: TimeInterval(nowMs) / 1000.0)
        let components = calendar.dateComponents([.hour, .minute], from: nowDate)
        let nowMinutes = (components.hour ?? 0) * 60 + (components.minute ?? 0)

        if start == end { return true }
        if start < end {
            return nowMinutes >= start && nowMinutes < end
        }
        return nowMinutes >= start || nowMinutes < end
    }

    private static func parseClock(_ value: String) -> Int? {
        let nsValue = value as NSString
        let range = NSRange(location: 0, length: nsValue.length)
        guard let regex = clockRegex,
              let match = regex.firstMatch(in: value, options: [], range: range),
              match.numberOfRanges == 3 else {
            return nil
        }

        let hhRange = match.range(at: 1)
        let mmRange = match.range(at: 2)
        guard hhRange.location != NSNotFound, mmRange.location != NSNotFound else {
            return nil
        }

        let hh = Int(nsValue.substring(with: hhRange))
        let mm = Int(nsValue.substring(with: mmRange))
        guard let hh, let mm, (0...23).contains(hh), (0...59).contains(mm) else {
            return nil
        }
        return (hh * 60) + mm
    }

    private static func scheduleKind(_ schedule: [String: Any]?) -> String {
        (schedule?["kind"] as? String) ?? ""
    }

    private static func int64(from value: Any?) -> Int64? {
        switch value {
        case nil:
            return nil
        case let n as Int64:
            return n
        case let n as Int:
            return Int64(n)
        case let n as Double:
            return Int64(n)
        case let n as NSNumber:
            return n.int64Value
        case let s as String:
            return Int64(s)
        default:
            return nil
        }
    }

    private static func bool(from value: Any?) -> Bool? {
        switch value {
        case nil:
            return nil
        case let b as Bool:
            return b
        case let n as NSNumber:
            return n.boolValue
        case let s as String:
            let normalized = s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            if normalized == "true" { return true }
            if normalized == "false" { return false }
            return nil
        default:
            return nil
        }
    }
}

File: android/src/main/java/io/mobilecron/NativeJobEvaluator.kt
package io.mobilecron

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.os.BatteryManager
import org.json.JSONArray
import org.json.JSONObject
import java.util.Calendar
import java.util.TimeZone

object NativeJobEvaluator {
    private const val STORAGE_FILE = "CapacitorStorage"
    private const val STORAGE_KEY = "mobilecron:state"
    private val CLOCK_REGEX = Regex("""^(\d{2}):(\d{2})$""")

    fun evaluate(context: Context, source: String): List<JSONObject> {
        val prefs = context.getSharedPreferences(STORAGE_FILE, Context.MODE_PRIVATE)
        val raw = prefs.getString(STORAGE_KEY, null) ?: return emptyList()
        val state = runCatching { JSONObject(raw) }.getOrNull() ?: return emptyList()

        val jobs = state.optJSONArray("jobs") ?: JSONArray()
        val pendingEvents = state.optJSONArray("pendingNativeEvents") ?: JSONArray()
        val firedEvents = mutableListOf<JSONObject>()
        val now = System.currentTimeMillis()
        val paused = state.optBoolean("paused", false)
        var mutated = false

        for (index in 0 until jobs.length()) {
            val job = jobs.optJSONObject(index) ?: continue
            if (!job.optBoolean("enabled", false)) continue

            if (readLong(job.opt("nextDueAt")) == null) {
                val computed = computeNextDueAt(job.optJSONObject("schedule"), now)
                if (computed != null) {
                    job.put("nextDueAt", computed)
                    mutated = true
                }
            }

            val nextDueAt = readLong(job.opt("nextDueAt"))
            if (nextDueAt == null || nextDueAt > now) continue

            val schedule = job.optJSONObject("schedule")
            if (getSkipReason(context, job, paused, now) != null) {
                job.put("consecutiveSkips", job.optInt("consecutiveSkips", 0) + 1)
                job.put("updatedAt", now)
                if (schedule?.optString("kind") == "every") {
                    val next = computeNextDueAt(schedule, now)
                    if (next != null) {
                        job.put("nextDueAt", next)
                    } else {
                        job.remove("nextDueAt")
                    }
                }
                mutated = true
                continue
            }

            job.put("lastFiredAt", now)
            job.put("updatedAt", now)
            job.put("consecutiveSkips", 0)

            val event = JSONObject().apply {
                put("id", job.optString("id"))
                put("name", job.optString("name"))
                put("firedAt", now)
                put("source", source)
                val data = job.opt("data")
                if (data != null && data != JSONObject.NULL) {
                    put("data", data)
                }
            }

            if (schedule?.optString("kind") == "at") {
                job.put("enabled", false)
                job.remove("nextDueAt")
            } else {
                val next = computeNextDueAt(schedule, now)
                if (next != null) {
                    job.put("nextDueAt", next)
                } else {
                    job.remove("nextDueAt")
                }
            }

            pendingEvents.put(event)
            firedEvents.add(event)
            mutated = true
        }

        if (mutated) {
            state.put("jobs", jobs)
            if (pendingEvents.length() > 0) {
                state.put("pendingNativeEvents", pendingEvents)
            } else {
                state.remove("pendingNativeEvents")
            }
            prefs.edit().putString(STORAGE_KEY, state.toString()).apply()
        }

        return firedEvents
    }

    internal fun computeNextDueAt(schedule: JSONObject?, nowMs: Long): Long? {
        if (schedule == null) return null
        return when (schedule.optString("kind")) {
            "at" -> {
                val atMs = readLong(schedule.opt("atMs")) ?: return null
                if (atMs > nowMs) atMs else null
            }
            "every" -> {
                val everyMs = readLong(schedule.opt("everyMs")) ?: return null
                if (everyMs <= 0) return null

                val anchorMs = readLong(schedule.opt("anchorMs")) ?: nowMs
                if (nowMs < anchorMs) return anchorMs

                val elapsed = nowMs - anchorMs
                val steps = (elapsed / everyMs) + 1
                anchorMs + (steps * everyMs)
            }
            else -> null
        }
    }

    internal fun isWithinActiveHours(activeHours: JSONObject, nowMs: Long): Boolean {
        val start = parseClock(activeHours.optString("start"))
        val end = parseClock(activeHours.optString("end"))
        if (start == null || end == null) return true

        val tzId = activeHours.optString("tz").takeIf { it.isNotBlank() }
        val timeZone = if (tzId != null) TimeZone.getTimeZone(tzId) else TimeZone.getDefault()
        val calendar = Calendar.getInstance(timeZone).apply { timeInMillis = nowMs }
        val nowMinutes = calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)

        if (start == end) return true
        if (start < end) {
            return nowMinutes >= start && nowMinutes < end
        }
        return nowMinutes >= start || nowMinutes < end
    }

    internal fun parseClock(value: String): Int? {
        val match = CLOCK_REGEX.matchEntire(value) ?: return null
        val hh = match.groupValues[1].toIntOrNull() ?: return null
        val mm = match.groupValues[2].toIntOrNull() ?: return null
        if (hh !in 0..23 || mm !in 0..59) return null
        return hh * 60 + mm
    }

    private fun getSkipReason(context: Context, job: JSONObject, paused: Boolean, nowMs: Long): String? {
        if (paused) return "paused"

        val activeHours = job.optJSONObject("activeHours")
        if (activeHours != null && !isWithinActiveHours(activeHours, nowMs)) {
            return "outside_active_hours"
        }

        if (job.optBoolean("requiresNetwork", false) && !isNetworkAvailable(context)) {
            return "requires_network"
        }

        if (job.optBoolean("requiresCharging", false) && !isChargingAvailable(context)) {
            return "requires_charging"
        }

        return null
    }

    private fun isNetworkAvailable(context: Context): Boolean {
        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager ?: return true
        return try {
            manager.activeNetworkInfo?.isConnected == true
        } catch (_: SecurityException) {
            true
        }
    }

    private fun isChargingAvailable(context: Context): Boolean {
        val batteryIntent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED)) ?: return false
        val status = batteryIntent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
        return status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL
    }

    private fun readLong(value: Any?): Long? {
        return when (value) {
            null, JSONObject.NULL -> null
            is Number -> value.toLong()
            is String -> value.toLongOrNull()
            else -> null
        }
    }
}
